doc/man/man4/cmot.4.man
'\" rtp
.\" vim: ft=nroff sw=4 noet nocin nosi com=b\:.\\\" fo+=tcqlorn tw=77
.\" =========================================================================
.\"
.\" @(#) doc/man/man4/cmot.4.man
.\"
.\" =========================================================================
.\"
.\" Copyright (c) 2008-2011  Monavacon Limited <http://www.monavacon.com/>
.\" Copyright (c) 2001-2008  OpenSS7 Corporation <http://www.openss7.com/>
.\" Copyright (c) 1997-2001  Brian F. G. Bidulock <bidulock@openss7.org>
.\"
.\" All Rights Reserved.
.\"
.\" Permission is granted to copy, distribute and/or modify this manual under
.\" the terms of the GNU Free Documentation License, Version 1.3 or any later
.\" version published by the Free Software Foundation; with no Invariant
.\" Sections, no Front-Cover Texts, and no Back-Cover Texts.  A copy of the
.\" license is included in the section entitled "GNU Free Documentation
.\" License".
.\"
.\" Permission to use, copy and distribute this manual without modification,
.\" for any purpose and without fee or royalty is hereby granted, provided
.\" that both the above copyright notice and this permission notice appears
.\" in all copies and that the name of OpenSS7 Corporation not be used in
.\" advertising or publicity pertaining to distribution of this documentation
.\" or its contents without specific, written prior permission.  OpenSS7
.\" Corporation makes no representation about the suitability of this manual
.\" for any purpose.  It is provided "as is" without express or implied
.\" warranty.
.\"
.\" Permission is granted to process this file through groff and print the
.\" results, provided the printed document carries a copying permission
.\" notice identical to this one except for the removal of this paragraph
.\" (this paragraph not being relevant to the printed manual).
.\"
.\" OPENSS7 CORPORATION DISCLAIMS ALL WARRANTIES WITH REGARD TO THIS MANUAL
.\" INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
.\" PARTICULAR PURPOSE, NON-INFRINGEMENT, OR TITLE; THAT THE CONTENTS OF THE
.\" DOCUMENT ARE SUITABLE FOR ANY PURPOSE, OR THAT THE IMPLEMENTATION OF SUCH
.\" CONTENTS WILL NOT INFRINGE ON ANY THIRD PARTY PATENTS, COPYRIGHTS,
.\" TRADEMARKS OR OTHER RIGHTS.  IN NO EVENT SHALL OPENSS7 CORPORATION BE
.\" LIABLE FOR ANY DIRECT, INDIRECT, SPECIAL OR CONSEQUENTIAL DAMAGES OR ANY
.\" DAMAGES WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER
.\" IN AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING
.\" OUT OF OR IN CONNECTION WITH ANY USE OF THIS DOCUMENT OR THE PERFORMANCE
.\" OR IMPLEMENTATION OF THE CONTENTS THEREOF.
.\" 
.\" Since the Linux kernel and libraries are constantly changing, this manual
.\" page may be incorrect or out-of-date.  The author(s) assume no
.\" responsibility for errors or omissions, or for damages resulting from the
.\" use of the information contained herein.  The author(s) may not have
.\" taken the same level of care in the production of this manual, which is
.\" licensed free of charge, as they might when working professionally.  The
.\" author(s) will take no responsibility in it.
.\" 
.\" Formatted or processed versions of this manual, if unaccompanied by the
.\" source, must acknowledge the copyright and authors of this work.
.\"
.\" -------------------------------------------------------------------------
.\"
.\" U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software on
.\" behalf of the U.S. Government ("Government"), the following provisions
.\" apply to you.  If the Software is supplied by the Department of Defense
.\" ("DoD"), it is classified as "Commercial Computer Software" under
.\" paragraph 252.227-7014 of the DoD Supplement to the Federal Acquisition
.\" Regulations ("DFARS") (or any successor regulations) and the Government
.\" is acquiring only the license rights granted herein (the license rights
.\" customarily provided to non-Government users).  If the Software is
.\" supplied to any unit or agency of the Government other than DoD, it is
.\" classified as "Restricted Computer Software" and the Government's rights
.\" in the Software are defined in paragraph 52.227-19 of the Federal
.\" Acquisition Regulations ("FAR") (or any successor regulations) or, in the
.\" cases of NASA, in paragraph 18.52.227-86 of the NASA Supplement to the
.\" FAR (or any successor regulations).
.\"
.\" =========================================================================
.\" 
.\" Commercial licensing and support of this software is available from
.\" OpenSS7 Corporation at a fee.  See http://www.openss7.com/
.\" 
.\" =========================================================================
.\"
.R1
bracket-label "\fR[\fB" "\fR]" "\fR, \fB"
no-default-database
database openss7.refs
accumulate
move-punctuation
abbreviate A
join-authors ", " ", " " and "
et-al " et al" 2 3
abbreviate-label-ranges ".."
sort-adjacent-labels
search-truncate 40
search-ignore CGIQOSTU
.R2
.so openss7.macros
.\"
.\"
.TH CMOT 4 "@PACKAGE_DATE@" "@PACKAGE@-@VERSION@" "@PACKAGE_TITLE@ Devices"
.\"
.\"
.SH NAME
.B cmot
\- CMISE over TCP
.\"
.\"
.SH SYNOPSIS
.PP
.HP
.IB cmip_stream " = open(" \(lq/dev/cmot\(rq ", O_RDWR);
.\"
.\"
.SH DESCRIPTION
.PP
.B CMOT
is an implementation of CMIP over TCP as described in
.IR "ISO Management over IP (CMOT)" ,
RFC 1095,
.[
rfc1095
.]
superseded by
RFC 1189,
.[
rfc1189
.]
with assigned IANA port number 163 (cmip-man).
.\"
.\"
.SH USAGE
.PP
.B CMOT
is used for communicating with a CMIP agent or client over a TCP/IP network.
On the client side, a
.B /dev/cmot
device is opened and connected to the CMIP server using the presentation
address as described in RFC 1069.
.[
rfc1069
.]
Then operations and components are exchanged using the XTI-mOSI interface.
.PP
On the server side, the
.B /dev/cmot
device is opened and listens on the presentation service access point as
described in RFC 1069.
.[
rfc1069
.]
When a call indication comes in, the CMIP server opens another
.B /dev/cmot
Stream and accepts the connection using the XTI-mOSI interface and begins
servicing requests.
.\"
.\"
.SH INTERFACE
.PP
The interface presented is the Transport Provider Interface Revision 2.0.0,
.[
tpi revision 2.0.0 draft 2
.]
.BR tpi (7),
with the minimal OSI extensions described in XNS 5.2.
.[
xns98
.]
.\"
.\"
.SH ADDRESSES
.PP
Addresses are formatted as presentation addresses as described in RFC 986
.[
rfc986
.]
and superseded by RFC 1069.
.[
rfc1069
.]
.\"
.\"
.SH OIDS
.PP
The following values for Application Context Name are registered for CMOT and
CMIP:
.PP
This OBJECT IDENTIFIER value is for use with legacy or backward compatible CMOT
.[
rfc1095
.]
operation.
.HP
.RB \(lq \ex2b\ex6\ex1\ex2\ex1\ex9\ex1\ex1 \(rq
.br
.I { iso(1) identified-organization(3) dod(6) internet(1) mgmt(2) mib(1) oim(9) acn(1) cmot1095(1) }
.PP
These OBJECT IDENTIFIER values are for use with CMIP
.[
rfc1189
.]
operation:
.HP
.RB \(lq \ex2b\ex6\ex1\ex2\ex1\ex9\ex1\ex2 \(rq
.br
.I { iso(1) identified-organization(3) dod(6) internet(1) mgmt(2) mib(1) oim(9) acn(1) manager-event-association(2) }
.HP
.RB \(lq \ex2b\ex6\ex1\ex2\ex1\ex9\ex1\ex3 \(rq
.br
.I { iso(1) identified-organization(3) dod(6) internet(1) mgmt(2) mib(1) oim(9) acn(1) manager-event-monitor-association(3) }
.HP
.RB \(lq \ex2b\ex6\ex1\ex2\ex1\ex9\ex1\ex4 \(rq
.br
.I { iso(1) identified-organization(3) dod(6) internet(1) mgmt(2) mib(1) oim(9) acn(1) manager-monitor-control-association(4) }
.HP
.RB \(lq \ex2b\ex6\ex1\ex2\ex1\ex9\ex1\ex5 \(rq
.br
.I { iso(1) identified-organization(3) dod(6) internet(1) mgmt(2) mib(1) oim(9) acn(1) manager-full-association(5) }
.HP
.RB \(lq \ex2b\ex6\ex1\ex2\ex1\ex9\ex1\ex6 \(rq
.br
.I { iso(1) identified-organization(3) dod(6) internet(1) mgmt(2) mib(1) oim(9) acn(1) agent-event-association(6) }
.\"
.\"
.SH OPTIONS
.PP
Quality of Service paramters are as described in RFC 986
.[
rfc986
.]
and superseded by RFC 1069.
.[
rfc1069
.]
.\"
.\"
.SH PRIMITIVES
.PP
The primitives of the connection-oriented transport service provider.
.HP
\fBT_BIND_REQ\fR(7)
.PD 0
.HP
\fBT_BIND_ACK\fR(7)
.HP
\fBT_INFO_REQ\fR(7)
.HP
\fBT_INFO_ACK\fR(7)
.HP
\fBT_CAPABILITY_REQ\fR(7)
.HP
\fBT_CAPABILITY_ACK\fR(7)
.HP
\fBT_CONN_REQ\fR(7)
.HP
\fBT_CONN_IND\fR(7)
.HP
\fBT_CONN_RES\fR(7)
.HP
\fBT_CONN_CON\fR(7)
.HP
\fBT_OPTDATA_REQ\fR(7)
.HP
\fBT_OPTDATA_IND\fR(7)
.HP
\fBT_DATA_REQ\fR(7)
.HP
\fBT_DATA_IND\fR(7)
.HP
\fBT_EXDATA_REQ\fR(7)
.HP
\fBT_EXDATA_IND\fR(7)
.HP
\fBT_DISCON_REQ\fR(7)
.HP
\fBT_DISCON_IND\fR(7)
.HP
\fBT_ORDREL_REQ\fR(7)
.HP
\fBT_ORDREL_IND\fR(7)
.HP
\fBT_UNBIND_REQ\fR(7)
.HP
\fBT_OK_ACK\fR(7)
.HP
\fBT_ERROR_ACK\fR(7)
.PD
.\"
.\"
.SH IOCTLS
.PP
None.
.\"
.\"
.SH NOTICES
.PP
None of this has been implemented yet.
.\"
.\"
.SH IMPLEMENTATION
.PP
.B CMOT
requires the services of the LPP (Lightweight Presentation Protocol),
RFC 1085,
.[
rfc1085
.]
.BR lpp (4),
which can be implemented as a
.I STREAMS
pushable module that is pushed over a TCP (tpi) Stream to form a TPI-mOSI
Stream.
.PP
.B CMOT
can be implemented, also, as a
.I STREAMS
pushable module that is then pushed over the LPP (tpi-mosi) Stream, or it can be
implemented as a user-space library that simply uses the LPP (tpi-mosi)
Stream.
Note that CMISE
.[
x710
.]
.[
x711
.]
uses ACSE and could easily be represented using the TPI-mOSI
in the same fashion as is done for SS7 TCAP.
.\"
.\"
.SH FILES
.PP
Obviously there are some GDMO files somewhere.
.\"
.\"
.SH MODULES
.PP
.TP
.B cmot
the CMOT pushable module.
.TP
.B lpp
the LPP pushable module.
.\"
.\"
.SH DEVICES
.PP
.TP
.B /dev/cmot
the CMOT device, which is an
.BR lpp (4)
device with the CMOT module autopushed.
.TP
.B /dev/lpp
the
.BR lpp (4)
deivce which is a TCP device with the LPP module autopushed.
.\"
.\"
.SH "SEE ALSO"
.PP
.BR ns-tcp (4),
.BR xol (4),
.BR xot (4),
.BR isot (4),
.BR itot (4),
.BR iso-ip (4),
.BR tp4 (4),
.IR ISO (4).
.\"
.\"
.SH BUGS
.PP
None of this is implemented yet, so it cannot have any bugs yet.
.\"
.\"
.SH COMPATIBILITY
.PP
None.
.\"
.\"
.SH CONFORMANCE
.PP
Too many to mention just now.
.\"
.\"
.SH HISTORY
.PP
The modules and drivers listed here are new for
.BR "@PACKAGE_TITLE@" .
.\"
.\"
.[
$LIST$
.]
.TI


NEWS
OpenSS7 -- history of user visible changes.  2015-01-15
$Id: NEWS,v 1.1.2.25 2011-05-10 13:45:30 brian Exp $
Copyright (c) 2008-2015  Monavacon Limited. <http://www.monavacon.com/>
Copyright (c) 2001-2008  OpenSS7 Corporation. <http://www.openss7.com/>
Copyright (c) 1997-2001  Brian Bidulock <bidulock@openss7.org>
See the end for copying conditions (for this file).

The sections that follow provide information on OpenSS7 releases of the
'OpenSS7' package.

Major changes for release openss7-1.1.7.20150115
------------------------------------------------

This is the seventh release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Source code release on GitHub.

Major changes for release openss7-1.1.7.20131209
------------------------------------------------

This is the seventh release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Build corrections for RHEL 6.4.

   - Scripts for country boundary, shoreline and rivers database
     preparation.

   - Database for ANSI T1.101 point code assignments.

   - Suppress extransous /proc/1/comm error messages when testing for
     systemd.

   - Handle PUIAS to Springdale rename for PUIAS distribution.

Major changes for release openss7-1.1.7.20131123
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Correction to ca-cert handling in RHEL 6 install scriptlet.

   - RHEL 6 build changes.

   - Set permissions correctly when applying kernel module patch.

   - Do not duplicate certs when including ca-cert.

   - Avoid conflicting shelll variable in RPM install scriplet.

   - Remove /proc/1/comm checks for systemd in module loader.

Major changes for release openss7-1.1.7.20130209
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Corrections to RPM spec file.

Major changes for release openss7-1.1.7.20130129
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Compensate for RHEL6 lack of c_rehash.

   - Do not create sysfs nodes: they do not work properly.

   - Use no foreground flag in strace.service file.

   - More error traces for strerr logger.

   - Remove static device names (the do not work either).

   - Changes to support weak kernel modules.

   - Clean up weak module support.

Major changes for release openss7-1.1.7.20130125
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Better Archlinux install script support.

Major changes for release openss7-1.1.6.20130125
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Remove use of kill_litter_super().

Major changes for release openss7-1.1.3.20130125
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Updates and corrections to weak kernel module builds.

   - Repo support and corrections for Archlinux packages.

Major changes for release openss7-1.1.3.20130123
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Additional support for Archlinux.

   - Add udev rules for strace and strerr.

   - Add system unit files for strace and strerr.

   - Add repo support for Archlinux packages.

   - Build Archlinux packages better.

Major changes for release openss7-1.1.3.20130121
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Build support for SLES 11.

   - Handle certificates better.

   - Get configuration files in the right places for SLES 11.

   - Support in RPM spec file for certificates on SLES 11.

   - Create a new OpenSS7 certificate.

   - Repository support for SLES 11.

   - RPM spec file reworked for new RPM release.

   - Better certificate handling for RHEL.

   - Additional build support for RHEL 6 with certificates.


src/modules/cmot.c
/*****************************************************************************

 @(#) File: src/modules/cmot.c

 -----------------------------------------------------------------------------

 Copyright (c) 2008-2015  Monavacon Limited <http://www.monavacon.com/>
 Copyright (c) 2001-2008  OpenSS7 Corporation <http://www.openss7.com/>
 Copyright (c) 1997-2001  Brian F. G. Bidulock <bidulock@openss7.org>

 All Rights Reserved.

 This program is free software: you can redistribute it and/or modify it under
 the terms of the GNU Affero General Public License as published by the Free
 Software Foundation, version 3 of the license.

 This program is distributed in the hope that it will be useful, but WITHOUT
 ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 FOR A PARTICULAR PURPOSE.  See the GNU Affero General Public License for more
 details.

 You should have received a copy of the GNU Affero General Public License
 along with this program.  If not, see <http://www.gnu.org/licenses/>, or
 write to the Free Software Foundation, Inc., 675 Mass Ave, Cambridge, MA
 02139, USA.

 -----------------------------------------------------------------------------

 U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software on
 behalf of the U.S. Government ("Government"), the following provisions apply
 to you.  If the Software is supplied by the Department of Defense ("DoD"), it
 is classified as "Commercial Computer Software" under paragraph 252.227-7014
 of the DoD Supplement to the Federal Acquisition Regulations ("DFARS") (or any
 successor regulations) and the Government is acquiring only the license rights
 granted herein (the license rights customarily provided to non-Government
 users).  If the Software is supplied to any unit or agency of the Government
 other than DoD, it is classified as "Restricted Computer Software" and the
 Government's rights in the Software are defined in paragraph 52.227-19 of the
 Federal Acquisition Regulations ("FAR") (or any successor regulations) or, in
 the cases of NASA, in paragraph 18.52.227-86 of the NASA Supplement to the FAR
 (or any successor regulations).

 -----------------------------------------------------------------------------

 Commercial licensing and support of this software is available from OpenSS7
 Corporation at a fee.  See http://www.openss7.com/

 *****************************************************************************/

static char const ident[] = "src/modules/cmot.c (" PACKAGE_ENVR ") " PACKAGE_DATE;

/*
 *  CMISE over TCP (CMOT and CMIP).
 */

#define _SVR4_SOURCE    1
#define _MPS_SOURCE     1
#define _SUN_SOURCE     1

#include <sys/os7/compat.h>

/* ---------------------- */

#define CMOT_DESCRIP	"CMOT STREAMS Module"
#define CMOT_EXTRA      "Part of the OpenSS7 OSI Stack for Linux Fast-STREAMS"
#define CMOT_COPYRIGHT	"Copyright (c) 2008-2015  Monavacon Limited.  All Rights Reserved."
#define CMOT_REVISION	"OpenSS7 src/modules/cmot.c (" PACKAGE_ENVR ") " PACKAGE_DATE
#define CMOT_DEVICE	"SVR 4.2MP CMISE over TCP (CMOT)"
#define CMOT_CONTACT	"Brian Bidulock <bidulock@openss7.org>"
#define CMOT_LICENSE	"GPL"
#define CMOT_BANNER	CMOT_DESCRIP	"\n" \
			CMOT_EXTRA	"\n" \
			CMOT_COPYRIGHT	"\n" \
			CMOT_REVISION	"\n" \
			CMOT_DEVICE	"\n" \
			CMOT_CONTACT	"\n"
#define CMOT_SPLASH	CMOT_DESCRIP	" - " \
			CMOT_REVISION

#ifndef CONFIG_STREAMS_CMOT_NAME
#error "CONFIG_STREAMS_CMOT_NAME must be defined."
#endif				/* CONFIG_STREAMS_CMOT_NAME */
#ifndef CONFIG_STREAMS_CMOT_MODID
#error "CONFIG_STREAMS_CMOT_MODID must be defined."
#endif				/* CONFIG_STREAMS_CMOT_MODID */

#ifdef LINUX
#ifdef MODULE
MODULE_AUTHOR(CMOT_CONTACT);
MODULE_DESCRIPTION(CMOT_DESCRIP);
MODULE_SUPPORTED_DEVICE(CMOT_DEVICE);
#ifdef MODULE_LICENSE
MODULE_LICENSE(CMOT_LICENSE);
#endif				/* MODULE_LICENSE */
#ifdef MODULE_ALIAS
MODULE_ALIAS("streams-cmot");
MODULE_ALIAS("streams-module-cmot");
MODULE_ALIAS("streams-modid-" __stringify(CONFIG_STREAMS_CMOT_MODID));
#endif				/* MODULE_ALIAS */
#ifdef MODULE_VERSION
MODULE_VERSION(PACKAGE_ENVR);
#endif				/* MODULE_VERSION */
#endif				/* MODULE */
#endif				/* LINUX */

#define CMOT_MOD_ID	CONFIG_STREAMS_CMOT_MODID
#define CMOT_MOD_NAME	CONFIG_STREAMS_CMOT_NAME

#ifndef CMOT_MOD_NAME
#define CMOT_MOD_NAME	"cmot"
#endif				/* CMOT_MOD_NAME */

#ifndef CMOT_MOD_ID
#define CMOT_MOD_ID	0
#endif				/* CMOT_MOD_ID */

/*
 *  STREAMS DEFINITIONS
 *  -------------------------------------------------------------------------
 */

#define MOD_ID		CMOT_MOD_ID
#define MOD_NAME	CMOT_MOD_NAME
#ifdef MODULE
#define MOD_BANNER	CMOT_BANNER
#else				/* MODULE */
#define MOD_BANNER	CMOT_SPLASH
#endif				/* MODULE */

/*
 *  LINUX INITIALIZATION
 *  --------------------
 */


#ifdef LINUX
unsigned short modid = MOD_ID;

#ifndef module_param
MODULE_PARM(modid, "h");
#else				/* module_param */
module_param(modid, short, 0444);
#endif				/* module_param */
MODULE_PARM_DESC(modid, "Module ID for CMOT module.  (0 for allocation.)");

/** cmotinit - initialize CMOT
  */
static __init int
cmotinit(void)
{
	cmn_err(CE_NOTE, MOD_BANNER);
	return (0);
}

/** cmotexit - terminate CMOT
  */
static __exit void
cmotexit(void)
{
	return;
}

module_init(cmotinit);
module_exit(cmotexit);

#endif				/* LINUX */
